### PR TITLE
fix: export command

### PR DESCRIPTION
**Pull request by rootulp**
Closes https://github.com/celestiaorg/celestia-app/issues/3392
Opens https://github.com/celestiaorg/celestia-app/issues/3472

Fixes a few bugs:
1. Previously all modules had `ExportGenesis` invoked on them even if they weren't supported by the current app version. Now we only call `ExportGenesis` for the modules that are supported by the current app version
2. The export command wasn't updated to account for the changes in https://github.com/celestiaorg/celestia-app/pull/3320 which force us to mount stores after `app.New()` based on the current app version
3. The minfee module couldn't be exported b/c it didn't register a key table in `ExportGenesis`

## Testing

I could export an app on app version 1 and 2. See [output](https://gist.github.com/rootulp/dfea2b5b40f7366b03706fc39321ceee)

### CODE REVIEW

**Comment by rootulp** [comment]
Why is the `Info` command reseting baseapp.AppVersion from 2 -> 1?!

File: app/export.go [hunk 1]
```diff
@@ -4,28 +4,32 @@ import (
...
+// ExportAppStateAndValidators exports the state of the application for a
+// genesis file.
+func (app *App) ExportAppStateAndValidators(forZeroHeight bool, jailAllowedAddrs []string) (servertypes.ExportedApp, error) {
+	ctx, err := app.CreateQueryContext(app.LastBlockHeight(), false)
+	if err != nil {
+		return servertypes.ExportedApp{}, err
+	}
+
+	app.InitializeAppVersion(ctx)
+	app.mountKeysAndInit(app.AppVersion())
```
Comment (rootulp) [inline app/export.go:23]: This is important because without this, we couldn't call `export` on the app because it wasn't fully initialized. A recent PR made it so that the app _needs_ to have it's app version initialized + key stores mounted _after_ `app.New` is invoked.

File: app/export.go [hunk 2]
```diff
@@ -36,12 +40,25 @@ func (app *App) ExportAppStateAndValidators(
...
 	validators, err := staking.WriteValidators(ctx, app.StakingKeeper)
+	if err != nil {
+		return servertypes.ExportedApp{}, err
+	}
+
 	return servertypes.ExportedApp{
 		AppState:        appState,
 		Validators:      validators,
-		Height:          height,
+		Height:          app.getExportHeight(forZeroHeight),
```
Comment (rootulp) [inline app/export.go:50]: [refactor] this extracts logic to a helper to make it easier to read

File: app/module/manager.go [hunk 3]
```diff
@@ -179,11 +179,17 @@ func (m *Manager) InitGenesis(ctx sdk.Context, cdc codec.JSONCodec, genesisData
...
 	genesisData := make(map[string]json.RawMessage)
 	modules := m.versionedModules[version]
-	for _, moduleName := range m.OrderExportGenesis {
+	moduleNamesForVersion := m.ModuleNames(version)
+	moduleNamesToExport := filter(m.OrderExportGenesis, func(moduleName string) bool {
+		// filter out modules that are not supported by this version
+		return slices.Contains(moduleNamesForVersion, moduleName)
+	})
+	for _, moduleName := range moduleNamesToExport {
 		genesisData[moduleName] = modules[moduleName].ExportGenesis(ctx, cdc)
```
Comment (rootulp) [inline app/module/manager.go:193]: this fixes a bug so that `ExportGenesis` is only invoked on the modules that are supported for the given `version`

File: app/module/manager_test.go [hunk 4]
```diff
@@ -146,34 +146,71 @@ func TestManager_InitGenesis(t *testing.T) {
...
+		mockAppModule1.EXPECT().ExportGenesis(gomock.Eq(ctx), gomock.Eq(cdc)).Times(1).Return(json.RawMessage(`{"key1": "value1"}`))
+		mockAppModule2.EXPECT().ExportGenesis(gomock.Eq(ctx), gomock.Eq(cdc)).Times(1).Return(json.RawMessage(`{"key2": "value2"}`))
+
+		want := map[string]json.RawMessage{
+			"module1": json.RawMessage(`{"key1": "value1"}`),
+			"module2": json.RawMessage(`{"key2": "value2"}`),
+		}
+		require.Equal(t, want, mm.ExportGenesis(ctx, cdc, 1))
+	})
+	t.Run("export genesis with one modules at version 1, one modules at version 2", func(t *testing.T) {
```
Comment (rootulp) [inline app/module/manager_test.go:179]: unit tests the fix

File: cmd/celestia-appd/cmd/root.go [hunk 5]
```diff
@@ -255,23 +254,22 @@ func NewAppServer(logger log.Logger, db dbm.DB, traceStore io.Writer, appOpts se
...
-		if err := capp.LoadHeight(height); err != nil {
+		if err := celestiaApp.LoadHeight(height); err != nil {
 			return servertypes.ExportedApp{}, err
 		}
-	} else {
-		capp = app.New(logger, db, traceStore, uint(1), encCfg, 0, appOpts)
 	}
-
-	return capp.ExportAppStateAndValidators(forZeroHeight, jailWhiteList)
+	return celestiaApp.ExportAppStateAndValidators(forZeroHeight, jailWhiteList)
```
Comment (rootulp) [inline cmd/celestia-appd/cmd/root.go:272]: refactored for readability

File: go.mod [hunk 6]
```diff
@@ -257,7 +257,7 @@ require (
 )
 
 replace (
-	github.com/cosmos/cosmos-sdk => github.com/celestiaorg/cosmos-sdk v1.22.0-sdk-v0.46.16
+	github.com/cosmos/cosmos-sdk => /Users/rootulp/git/rootulp/celestiaorg/cosmos-sdk
```
Comment (rootulp) [inline go.mod:260]: need to revert

**Comment by rootulp** [comment]
```
panic: parameter GlobalMinGasPrice not registered

goroutine 1 [running]:
github.com/cosmos/cosmos-sdk/x/params/types.Subspace.checkType({{0x105127140, 0x14000122c20}, 0x140013b6060, {0x1050fb1e8, 0x14000122db0}, {0x1050fb210, 0x14000123110}, {0x14001b8e208, 0x6, 0x8}, ...}, ...)
	/Users/rootulp/git/rootulp/celestiaorg/cosmos-sdk/x/params/types/subspace.go:170 +0x174
github.com/cosmos/cosmos-sdk/x/params/types.Subspace.Get({{_, _}, _, {_, _}, {_, _}, {_, _, _}, ...}, ...)
	/Users/rootulp/git/rootulp/celestiaorg/cosmos-sdk/x/params/types/subspace.go:104 +0x80
github.com/celestiaorg/celestia-app/v2/x/minfee.ExportGenesis({{0x105115438, 0x106557920}, {0x105127840, 0x140010b3d80}, {{0x0, 0x0}, {0x0, 0x0}, 0x4, {0x0, ...}, ...}, ...}, ...)
	/Users/rootulp/git/rootulp/celestiaorg/celestia-app/x/minfee/genesis.go:31 +0x1e0
github.com/celestiaorg/celestia-app/v2/x/minfee.AppModule.ExportGenesis({{}, {{_, _}, _, {_, _}, {_, _}, _}}, {{0x105115438, ...}, ...}, ...)
	/Users/rootulp/git/rootulp/celestiaorg/celestia-app/x/minfee/module.go:132 +0x84
github.com/celestiaorg/celestia-app/v2/app/module.(*Manager).ExportGenesis(_, {{0x105115438, 0x106557920}, {0x105127840, 0x140010b3d80}, {{0x0, 0x0}, {0x0, 0x0}, 0x4, ...}, ...}, ...)
	/Users/rootulp/git/rootulp/celestiaorg/celestia-app/app/module/manager.go:193 +0x344
github.com/celestiaorg/celestia-app/v2/app.(*App).ExportAppStateAndValidators(0x1400079a008, 0x0, {0x106557920, 0x0, 0x0})
	/Users/rootulp/git/rootulp/celestiaorg/celestia-app/app/export.go:36 +0x334
github.com/celestiaorg/celestia-app/v2/cmd/celestia-appd/cmd.createAppAndExport({0x1051157f0, 0x1400133efc0}, {0x105127f18, 0x140013b6058}, {0x0, 0x0}, 0xffffffffffffffff, 0x0, {0x106557920, 0x0, ...}, ...)
	/Users/rootulp/git/rootulp/celestiaorg/celestia-app/cmd/celestia-appd/cmd/root.go:272 +0x158
github.com/cosmos/cosmos-sdk/server.ExportCmd.func1(0x14001346908, {0x106557920?, 0x0?, 0x0?})
	/Users/rootulp/git/rootulp/celestiaorg/cosmos-sdk/server/export.go:70 +0x3cc
github.com/spf13/cobra.(*Command).execute(0x14001346908, {0x106557920, 0x0, 0x0})
	/Users/rootulp/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:983 +0x840
github.com/spf13/cobra.(*Command).ExecuteC(0x14001283b08)
	/Users/rootulp/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1115 +0x344
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/rootulp/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1039
github.com/spf13/cobra.(*Command).ExecuteContext(...)
	/Users/rootulp/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1032
github.com/cosmos/cosmos-sdk/server/cmd.Execute(0x14001283b08, {0x103ee27a4, 0x8}, {0x140006f9f00, 0x1c})
	/Users/rootulp/git/rootulp/celestiaorg/cosmos-sdk/server/cmd/execute.go:36 +0x158
main.main()
	/Users/rootulp/git/rootulp/celestiaorg/celestia-app/cmd/celestia-appd/main.go:14 +0x3c
```

**Comment by rootulp** [comment]
Strange because InitGenesis is called on the minfee module when upgrading from app version 1 -> 2. The param does exist in the correct subspace:

```
$ celestia-appd query params subspace minfee GlobalMinGasPrice
key: GlobalMinGasPrice
subspace: minfee
value: '"0.000001000000000000"'
```

It looks like the export command isn't loading the correct params

File: app/export.go [hunk 7]
```diff
@@ -4,28 +4,34 @@ import (
...
+	if !app.IsSealed() {
+		app.mountKeysAndInit(app.AppVersion())
+	}
+
+	// Create a new context so that the commit multi-store reflects the store
+	// key mounting performed above.
+	ctx, err = app.CreateQueryContext(app.LastBlockHeight(), false)
+	if err != nil {
+		return servertypes.ExportedApp{}, err
+	}
```
Comment (rach-id) [inline app/export.go:32]: [nit]
If this should only be executed if `!app.IsSealed()`, then I think it makes sense for it to be inside the same block

File: app/test/export_test.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+		forZeroHeight := false
+		jailAllowedAddrs := []string{}
+
+		testApp, _ := SetupTestAppWithUpgradeHeight(t, 3)
+		upgradeToV2(t, testApp)
+
+		exported, err := testApp.ExportAppStateAndValidators(forZeroHeight, jailAllowedAddrs)
+		require.NoError(t, err)
+		assert.NotNil(t, exported)
+		// TODO: the following assertion is commented out because the exported app does not populate consensus params.version
```
Comment (rach-id) [inline app/test/export_test.go:35]: should we create an issue for this? or it's fine not to support exporting the version?

File: cmd/celestia-appd/cmd/root.go [hunk 9]
```diff
@@ -255,23 +254,22 @@ func NewAppServer(logger log.Logger, db dbm.DB, traceStore io.Writer, appOpts se
...
+	logger log.Logger,
+	db dbm.DB,
+	traceStore io.Writer,
+	height int64,
+	forZeroHeight bool,
+	jailWhiteList []string,
 	appOpts servertypes.AppOptions,
 ) (servertypes.ExportedApp, error) {
-	encCfg := encoding.MakeConfig(app.ModuleEncodingRegisters...) // Ideally, we would reuse the one created by NewRootCmd.
-	encCfg.Codec = codec.NewProtoCodec(encCfg.InterfaceRegistry)
```
Comment (rach-id) [inline cmd/celestia-appd/cmd/root.go:262]: [blocking]
why is the interface registry no longer registered?

File: scripts/single-node.sh [hunk 10]
```diff
@@ -82,13 +82,20 @@ sed -i'.bak' 's#"null"#"kv"#g' "${CELESTIA_APP_HOME}"/config/config.toml
 # Persist ABCI responses
 sed -i'.bak' 's#discard_abci_responses = true#discard_abci_responses = false#g' "${CELESTIA_APP_HOME}"/config/config.toml
 
+# Override the log level to debug
+sed -i'.bak' 's#log_level = "info"#log_level = "debug"#g' "${CELESTIA_APP_HOME}"/config/config.toml
```
Comment (rach-id) [inline scripts/single-node.sh:86]: [personal opinion][not blocking]
wouldn't this be too much for someone just wanting to run a node? debug logs are hard to follow

File: x/minfee/params.go [hunk 11]
```diff
@@ -29,11 +29,13 @@ type Params struct {
...
-func RegisterMinFeeParamTable(ps paramtypes.Subspace) {
-	if !ps.HasKeyTable() {
-		ps.WithKeyTable(ParamKeyTable())
+// RegisterMinFeeParamTable attaches a key table to the provided subspace if it
+// doesn't have one.
+func RegisterMinFeeParamTable(subspace paramtypes.Subspace) paramtypes.Subspace {
+	if !subspace.HasKeyTable() {
+		return subspace.WithKeyTable(ParamKeyTable())
 	}
+	return subspace
```
Comment (rach-id) [inline x/minfee/params.go:38]: [blocking]
Does the method make changes directly on the `subspace` or it is copied and needs to be returned. If it's the latter, then this also should change:

https://github.com/celestiaorg/celestia-app/blob/ef67d05690d445758472b50ecacaf8a5d5f2b0f2/test/tokenfilter/setup.go#L247

**Review by rach-id — COMMENTED** [review]
LGTM 👍 

Left two blocking comments to understand more

File: app/app.go [hunk 12]
```diff
@@ -169,11 +169,12 @@ type App struct {
...
 
-// New returns a reference to an initialized celestia app.
+// New returns a reference to an uninitialized app. Callers must subsequently
+// call app.Info or app.InitChain to initialize the baseapp.
 //
-// NOTE: upgradeHeight refers specifically to the height that
-// a node will upgrade from v1 to v2. It will be deprecated in v3
-// in place for a dynamically signalling scheme
+// NOTE: upgradeHeightV2 refers specifically to the height that a node will
+// upgrade from v1 to v2. It will be deprecated in v3 in place for a dynamically
```
[the remark this answers is not in the document]
Comment (rootulp) [reply app/app.go:176]: Yea there is a proposal and it's one of the features we're shipping in v2. 

https://github.com/celestiaorg/CIPs/blob/main/cips/cip-10.md

File: app/export.go (hunk 7 above)
Comment (rootulp) [reply app/export.go:32]: I think we want a new context regardless that accounts for the initialized app version (line 22) + the key mounting if it happened (line 24). 

Put another way, we want it outside the conditional

File: app/test/export_test.go (hunk 8 above)
Comment (rootulp) [reply app/test/export_test.go:35]: Already an issue https://github.com/celestiaorg/celestia-app/issues/3472

File: cmd/celestia-appd/cmd/root.go (hunk 9 above)
Comment (rootulp) [reply cmd/celestia-appd/cmd/root.go:262]: it's already registered by the previous line because `encoding.MakeConfig` does it

https://github.com/rootulp/celestia-app/blob/413befd6ade08e38973aa051f3f3814201342853/app/encoding/encoding.go#L41

File: scripts/single-node.sh (hunk 10 above)
Comment (rootulp) [reply scripts/single-node.sh:86]: I found them useful for debugging but I can understand not needing them on. Can revert.

My opinion is that this `single-node.sh` script is for developers / maintainers who are actively working on celestia-app so the configs in this file should be tailored to them. For example, tx indexing should be enabled even though it takes up more disk space b/c it's useful to query transactions.

I can see debug logs being noisy for other maintainers though so will revert

File: x/minfee/params.go (hunk 11 above)
Comment (rootulp) [reply x/minfee/params.go]: can do.

Comment (rootulp) [reply x/minfee/params.go]: 825f9263ddd54b62a0fb6c1f7ec1f09d0dfd561e

File: scripts/single-node.sh (hunk 10 above)
Comment (rootulp) [reply scripts/single-node.sh]: 2af00333831315af11cbb2c2a5bd2e6728bb4c0d

**Review by rach-id — APPROVED** [review]

**Review by evan-forbes — APPROVED** [review]
LGTM

File: app/test/export_test.go (hunk 8 above)
Comment (cmwaters) [reply app/test/export_test.go:35]: Have you tested to make sure that a user can export the state and start up a new node from that exported state i.e. I would assume not saving the consensus params.version would cause issues upon restart

File: app/module/manager.go (hunk 3 above)
Comment (cmwaters) [reply app/module/manager.go:193]: make sense

**Review by cmwaters — COMMENTED** [review]
Looks good to me. Nice find

File: app/test/export_test.go (hunk 8 above)
Comment (rootulp) [reply app/test/export_test.go:35]: I haven't tested but agreed it could cause issues so I expect a user would have to manually populate the app version prior to start.
